### DISCUSSION TITLE
Vertex Classification

### DISCUSSION THREAD

**Discussion by AbdBarho** [e3nn/e3nn#365] [Q&A]
Hi!

Thanks a lot for this great library. I am trying to use it and faced many questions, some of them perhaps stem from my lack of deep math knowledge, and I was hoping to get some confirmation that my thought process is correct, or at least guidance on what to research next.

Most of the questions are a simple yes/no!

### Problem
I wanted to try and use `e3nn` for doing point classification, more precisely, vertex classification on a 3d mesh.

The problem definition: given a 3d mesh (points + edges), train a model that predicts class probability per vertex. The meshes I have are of the human heart, always centered on `(0,0,0)`, consistently scaled, and are never mirrored, but always rotated differently.

**Question 1**: Put in symmetry terms, I need a model that produces a rotationally invariant scalar per vertex. Translation and reflection symmetry is not required. Is this correct? I get the other two "for free" with `e3nn`, does this hold if the output should be *invariant* ? (I assume yes).

### Implementation

Digging into the examples provided in this repo, the closest thing to my problem would be the [`Simple Network` class in `nn/models/v2106/gate_points_networks.py`](https://github.com/e3nn/e3nn/blob/19d0377da7925b684bade615ecefbc12627076a2/e3nn/nn/models/v2106/gate_points_networks.py#L29).

I created a simple network using the following parameters:
```python
irreps_in='1o'
irreps_out='0o'
```

**Question 2**: since the only input I have are the vertex coordinates (xyz), my `irreps_in` should only consist of a term with $L=1$, hence `irreps_in='1o'`, correct? does a point count as a vector in this case? I would basically call the network with
```python
net({
  'pos': pos,
  'x': pos
})
```

**Question 3**: since the output should be invariant, we only need a scalar, hence $L=0$ or `irreps_out='0e'`, correct? I am unsure about the parity in this context

I was hoping that if I could get my code up & running I could contribute it to the repo as an example.

I still have a lot of questions, but maybe this is enough to guide me through the rest. Also I am still going through the resources at https://e3nn.org/e3nn-tutorial-mrs-fall-2021/

Thanks again!

**Comment by mariogeiger** [comment]
**Answer 1** yes, you want your output to be a scalar (`irreps=0e`)

**Answer 2** in the network you use, the position is not an option and therefore is not part of `irreps_in`. In your application, the input nodes have no data. However you need to provide something to the network. The solution is to provides ones: `irreps_in=0e`

```
net({
  'pos': pos,
  'x': torch.ones(pos.shape[0], 1)
})
```

**Answer 3** yes `irreps_out=0e`. About parity, `Simple Network` is E(3) equivariant, you have no choice with this network. There is a way to create a SE(3) equivariant network but I don't think you need it: I think it is fair to assume that an inverted heart has the same node predictions.
